src/bgw/job_stat.c

```diff
@@ -0,0 +1,345 @@
+		 * have happened before failure)
+		 */
+		if (!bgw_job_stat_next_start_was_set(fd))
+			fd->next_start = calculate_next_start_failure(fd->last_finish, fd->consecutive_failures, res->job);
+	}
+
+	catalog_update(ti->scanrel, tuple);
+	heap_freetuple(tuple);
+	return false;
+}
+
+static bool
+bgw_job_stat_tuple_set_next_start(TupleInfo *ti, void *const data)
+{
+	TimestampTz *next_start = data;
+	HeapTuple	tuple = heap_copytuple(ti->tuple);
+	FormData_bgw_job_stat *fd = (FormData_bgw_job_stat *) GETSTRUCT(tuple);
+
+	fd->next_start = *next_start;
+
+	catalog_update(ti->scanrel, tuple);
+	heap_freetuple(tuple);
+	return false;
+}
+
+static bool
+bgw_job_stat_insert_mark_start_relation(Relation rel,
+										int32 bgw_job_id)
+{
+	TupleDesc	desc = RelationGetDescr(rel);
+	Datum		values[Natts_bgw_job_stat];
+	bool		nulls[Natts_bgw_job_stat] = {false};
+	CatalogSecurityContext sec_ctx;
+
+	Interval	zero_ival = {.time = 0,};
+
+	values[Anum_bgw_job_stat_job_id - 1] = Int32GetDatum(bgw_job_id);
+	values[Anum_bgw_job_stat_last_start - 1] = TimestampGetDatum(GetCurrentTimestamp());
+	values[Anum_bgw_job_stat_last_finish - 1] = TimestampGetDatum(DT_NOBEGIN);
+	values[Anum_bgw_job_stat_next_start - 1] = TimestampGetDatum(DT_NOBEGIN);
+	values[Anum_bgw_job_stat_total_runs - 1] = Int64GetDatum(1);
+	values[Anum_bgw_job_stat_total_duration - 1] = IntervalPGetDatum(&zero_ival);
+	values[Anum_bgw_job_stat_total_success - 1] = Int64GetDatum(0);
+	values[Anum_bgw_job_stat_total_failures - 1] = Int64GetDatum(0);
+	values[Anum_bgw_job_stat_consecutive_failures - 1] = Int32GetDatum(0);
+
+	/* This is udone by any of the end marks */
+	values[Anum_bgw_job_stat_last_run_success - 1] = BoolGetDatum(false);
+	values[Anum_bgw_job_stat_total_crashes - 1] = Int64GetDatum(1);
+	values[Anum_bgw_job_stat_consecutive_crashes - 1] = Int32GetDatum(1);
+
+	catalog_become_owner(catalog_get(), &sec_ctx);
+	catalog_insert_values(rel, desc, values, nulls);
+	catalog_restore_user(&sec_ctx);
+
+	return true;
+}
+
+static bool
+bgw_job_stat_insert_mark_start(int32 bgw_job_id)
+{
+	Catalog    *catalog = catalog_get();
+	Relation	rel;
+	bool		result;
+
+	rel = heap_open(catalog->tables[BGW_JOB_STAT].id, RowExclusiveLock);
+	result = bgw_job_stat_insert_mark_start_relation(rel, bgw_job_id);
+	heap_close(rel, RowExclusiveLock);
+
+	return result;
+}
+
+
+void
+ts_bgw_job_stat_mark_start(int32 bgw_job_id)
+{
+	if (bgw_job_stat_scan_job_id(bgw_job_id, bgw_job_stat_tuple_mark_start, NULL, NULL, RowExclusiveLock) == 0)
+		bgw_job_stat_insert_mark_start(bgw_job_id);
+
+}
+
+void
+ts_bgw_job_stat_mark_end(BgwJob *job, bool success)
+{
+	JobResult	res =
+	{
+		.job = job,
+		.success = success,
+	};
+
+	if (bgw_job_stat_scan_job_id(job->fd.id, bgw_job_stat_tuple_mark_end, NULL, &res, RowExclusiveLock) == 0)
+		elog(ERROR, "unable to find job statistics for job %d", job->fd.id);
+
+}
+
+bool
+ts_bgw_job_stat_end_was_marked(BgwJobStat *jobstat)
```

Comment:
This function name looks like it should be a getter for `ts_bgw_job_stat_mark_end`, maybe rename to something like `ts_bgw_job_stat_has_finish_timestamp`.

Reply:
I actually meant it as a getter for whether or not the "mark_end" operation was performed. Is there a better name you can suggest?

Reply:
Best I can think of is `.._was_finished` or `..._finished_cleanly` or something descriptive of how it should be used like that.

(The problem is that we have a lot of things starting with `ts_bgw_job_stat_...` split across both the struct and the tuple, in this case there's a `mark_end` in both, and one based on the other, so it's not obvious when each getter is valid. We may just have to accept confusing names. either way, we should document what the function means.)